File: src/basic/alg/l239.cpp
/**
 * L239 滑动窗口最大值
 *
 *输入：nums = [1,3,-1,-3,5,3,6,7], k = 3
  输出：[3,3,5,5,6,7]
  解释：
  滑动窗口的位置                最大值
  ---------------               -----
  [1  3  -1] -3  5  3  6  7       3
   1 [3  -1  -3] 5  3  6  7       3
   1  3 [-1  -3  5] 3  6  7       5
   1  3  -1 [-3  5  3] 6  7       5
   1  3  -1  -3 [5  3  6] 7       6
   1  3  -1  -3  5 [3  6  7]      7
 *
 * 1. 与 最小k 个数类似，如用最大堆处理，但是无法区别最先开始的值
 * 2. 优先队列
 * a. 最大堆
 * b. 超出窗口的要移除
 */

#include "../util/myprint.h"
#include <set>
#include <queue>


vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    vector<int> res;
    if (nums.size() == 0 || k <= 0) {
        return res;
    }

    int n = nums.size();
    priority_queue<pair<int, int>> q;  // 大根堆

    // 预先k数
    for (int i=0; i<k; i++) {
        q.emplace(nums[i], i);  // &&&
    }

    res.push_back(q.top().first);

    for (int i=k; i<n; i++) {
        q.emplace(nums[i], i);  // &&&

        // 超出窗口的移除，比较巧妙. i-k定位到首节点 &&& <=
        while (q.top().second <= i - k) {
            q.pop();
        }

        res.push_back(q.top().first);
    }

    return res;
}

int main(int argc, char **argv) {
    vector<int> nums{1,3,-1,-3,5,3,6,7};
    auto res = maxSlidingWindow(nums, 3);
    show(res);

    return 0;
}

File: src/basic/alg/L239.java
package basic.alg;

import util.MyPrint;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * 滑动窗口最大值
 * 输入：nums = [1,3,-1,-3,5,3,6,7], k = 3
 * 输出：[3,3,5,5,6,7]
 * 解释：
 * 滑动窗口的位置                最大值
 * ---------------               -----
 * [1  3  -1] -3  5  3  6  7       3
 * 1 [3  -1  -3] 5  3  6  7       3
 * 1  3 [-1  -3  5] 3  6  7       5
 * 1  3  -1 [-3  5  3] 6  7       5
 * 1  3  -1  -3 [5  3  6] 7       6
 * 1  3  -1  -3  5 [3  6  7]      7
 *
 * 1.单调队列，单调递增
 * 动态维持平衡，队头总是最大值。
 */
public class L239 {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if (nums.length == 0 || k == 0) {
            return new int[]{};
        }

        MQueue q = new MQueue();   // 单调队列
        List<Integer> list = new ArrayList<>(); // res

        for (int i = 0; i < nums.length; i++) {
            if (i < k - 1) {
                q.push(nums[i]);
            } else {
                q.push(nums[i]);
                list.add(q.max());

                // 移除队头数据
                q.pop(nums[i - k + 1]);
            }
        }

        int n = list.size();
        int[] res = new int[n];

        // copy
        for (int i=0; i<n; i++) {
            res[i] = list.get(i);
        }

        return res;
    }

    public static void main(String[] args) {
        L239 l = new L239();
        var res = l.maxSlidingWindow(new int[]{1,3,-1,-3,5,3,6,7}, 3);
        MyPrint.show(res);
    }
}

/**
 * 单调队列; 小的元素都被弹出，队头元素最大
 */
class MQueue {
    private LinkedList<Integer> queue;

    public MQueue() {
        queue = new LinkedList<>();
    }

    public void push(int n) {
        // 5 -> 4 -> 3
        // 前面比其小的都删除，数组递减. 与单调栈逻辑类似
        while (!queue.isEmpty() && queue.getLast() < n) {   // &&& <
            queue.pollLast();
        }

        queue.addLast(n);
    }

    public void pop(int n) {
        if (n == queue.getFirst()) {
            queue.pollFirst();
        }
    }

    public int max() {
        return queue.getFirst();
    }
}